src/onnx/parse_reduce_op.cpp

```diff
@@ -31,66 +31,123 @@ namespace migraphx {
-    {
-        axes.clear();
-        auto&& attr_axes = info.attributes["axes"].ints();
-        axes             = std::vector<int64_t>(attr_axes.begin(), attr_axes.end());
-    }
+        auto constant_axes = parse_constant_axes(args, info);
 
-    bool noop_with_empty_axes = false;
-    if(contains(info.attributes, "noop_with_empty_axes"))
-    {
-        noop_with_empty_axes = static_cast<bool>(
-            parser.parse_value(info.attributes.at("noop_with_empty_axes")).at<int>());
-    }
+        int noop_with_empty_axes =
+            parse_attribute<int>("noop_with_empty_axes", parser, info).value_or(0);
 
-    // empty axes behavior
-    if(axes.empty())
-    {
-        if(noop_with_empty_axes)
+        int keep_dims = parse_attribute<int>("keepdims", parser, info).value_or(1);
+
+        std::vector<int64_t> all_axes(args.front()->get_shape().ndim());
+        std::iota(all_axes.begin(), all_axes.end(), 0);
+
+        // Handle axes attribute, constant input axes, and missing both attribute and input cases
+        if(constant_axes.has_value())
+        {
+            if(noop_with_empty_axes != 0 and constant_axes->empty())
+            {
+                return args[0];
+            }
+
+            if(noop_with_empty_axes == 0 and constant_axes->empty())
+            {
+                constant_axes = all_axes;
+            }
+
+            auto reduce =
+                info.add_instruction(make_op(op_name, {{"axes", *constant_axes}}), args[0]);
+            if(keep_dims == 0)
+            {
+                return info.add_instruction(make_op("squeeze", {{"axes", *constant_axes}}), reduce);
+            }
+            return reduce;
+        }
+
+        // Handle variable input axes
+        if(keep_dims == 0)
+        {
+            MIGRAPHX_THROW("Keepdims not supported with runtime provided axes");
+        }
+        // Empty axes attribute indicates to the operator to look for axes in the inputs
+        auto reduce_op = make_op(op_name, {{"axes", {}}});
+
+        if(noop_with_empty_axes != 0)
+        {
+            return info.add_instruction(reduce_op, args);
+        }
+
+        if(args[1]->get_shape().dynamic())
+        {
+            auto reduce_input_axes = info.add_instruction(reduce_op, args);
+            auto all_axes_lit      = info.add_literal(
+                literal{shape{shape::type_t::int64_type, {all_axes.size()}}, all_axes});
+            auto reduce_all_axes = info.add_instruction(reduce_op, args[0], all_axes_lit);
+            auto zero      = info.add_literal(literal{shape{shape::type_t::int64_type}, {0u}});
+            auto axes_size = info.add_instruction(make_op("dimensions_of", {{"end", 1}}), args[1]);
+            auto is_axes_empty = info.add_instruction(make_op("equal"), axes_size, zero);
+            auto is_axes_empty_bc =
+                info.add_instruction(make_op("multibroadcast"), is_axes_empty, reduce_all_axes);
+
+            return info.add_instruction(
```

Comment:
@pfultz2 I've implemented operator parsing in the fashion you recommended. The `where` here will work during parsing, due to dynamic output shape equivalence between branches. However, when evaluation time comes around, the output shapes will differ which causes `where` to throw an exception(demonstrated by this [test](https://github.com/ROCm/AMDMIGraphX/pull/2120/files#diff-d6b0a0ea0ed42a315a1748fdd544a06facad4a564d07ff588bb9233348db615eR47)).